web/packages/shared/components/LatencyDiagnostic/LatencyDiagnostic.tsx

```diff
@@ -23,70 +23,222 @@ import { TeleportGearIcon } from 'design/SVGIcon';
 
 import { MenuIcon } from 'shared/components/MenuAction';
 
-function colorForLatency(l: number): string {
-  if (l > 400) {
-    return 'dataVisualisation.tertiary.abbey';
+export const warnThreshold = 150;
+export const errorThreshold = 400;
+
+export enum latencyColor {
+  ok = 'dataVisualisation.tertiary.caribbean',
+  warn = 'dataVisualisation.tertiary.abbey',
+  error = 'dataVisualisation.tertiary.sunflower',
+}
+
+// latencyColors determines the color to use for each leg of the connection
+// and the total measurement. The total color is derived from each leg of
+// the connection instead of the combined latency to prevent two latencies
+// in the same threshold from causing the total to jump into the next threshold.
+// For example, this prevents showing a warning if both legs of the connection are ok,
+// but their sum would result in exceeding the warning threshold.
+export function latencyColors(
+  client: number,
+  server: number
+): { client: latencyColor; server: latencyColor; total: latencyColor } {
+  function colorForLatency(l: number): latencyColor {
+    if (l >= errorThreshold) {
+      return latencyColor.error;
+    }
+
+    if (l >= warnThreshold) {
+      return latencyColor.warn;
+    }
+
+    return latencyColor.ok;
+  }
+
+  const clientColor = colorForLatency(client);
+  const serverColor = colorForLatency(server);
+
+  // any + red = red
+  if (client >= errorThreshold || server >= errorThreshold) {
+    return {
+      client: clientColor,
+      server: serverColor,
+      total: latencyColor.error,
+    };
+  }
+
+  // yellow + yellow = yellow
+  if (client >= warnThreshold && server >= warnThreshold) {
+    return {
+      client: clientColor,
+      server: serverColor,
+      total: latencyColor.warn,
+    };
   }
 
-  if (l > 150) {
-    return 'dataVisualisation.tertiary.sunflower';
+  // green + yellow = yellow
+  if (
+    (client >= warnThreshold || server >= warnThreshold) &&
+    (client < warnThreshold || server < warnThreshold)
+  ) {
```

Comment:
This is fine but it seems like (correct me if I'm wrong) that the pattern is basically "show the highest threshold as the total". So if any are red, it'll be red, and if any are yellow, it'll show yellow.

I wonder if instead of doing these comparisons we could just do another small helper function like `getColorForTotal` and have the logic be

if either are red, return red
if either are yellow, return yellow
return green

thoughts?